### Week 3 Linear Regression

In [1]:
# This code appears in every demonstration Notebook.
# By default, when you run each cell, only the last output of the codes will show.
# This code makes all outputs of a cell show.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

We will explore the user and order data from JD.com.

1. We import the necessary packages.

In [2]:
import pandas as pd
import statsmodels.api as sm
# api submodule gives access to the most commonly used 
# classes and functions directly.

2. We read in the datasets.

In [3]:
ads = pd.read_csv('Advertising (1).csv')
# if we use absolute oath
#ads =pd.read_csv("\\filepath ")

In [4]:
ads.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [5]:
# Delete a column
ads.columns


Index(['Unnamed: 0', 'TV', 'radio', 'newspaper', 'sales'], dtype='object')

In [6]:

del ads['Unnamed: 0']

In [7]:
ads.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

3. Run the linear regression model

In [36]:
# prepare the variables:dependent,independent and the constant
y = ads['sales']

X = ads[['TV', 'radio', 'newspaper']]

X_with_intercept = sm.add_constant(X)

# Fit the linear regression model ols means ordinatry list square
model = sm.OLS(y, X_with_intercept)
results = model.fit()

In [9]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Mon, 19 Feb 2024   Prob (F-statistic):           1.58e-96
Time:                        14:47:02   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [37]:
#make predictions #use predict() to make predictions
sales_predicted = results.predict(X_with_intercept)

In [48]:
ads['sales_predicted']= sales_predicted
ads.head()

TV  radio  newspaper  sales  sales_predicted
0  230.1   37.8       69.2   22.1        20.523974
1   44.5   39.3       45.1   10.4        12.337855
2   17.2   45.9       69.3    9.3        12.307671
3  151.5   41.3       58.5   18.5        17.597830
4  180.8   10.8       58.4   12.9        13.188672

In [72]:
import sklearn
import math

from sklearn.metrics import mean_squared_error
mean_squared_error

<function sklearn.metrics._regression.mean_squared_error(y_true, y_pred, *, sample_weight=None, multioutput='uniform_average', squared=True)>

In [59]:
#to get the number of 
ads.shape[0]

200

In [69]:
mean_squared_error(ads['sales'],ads['sales_predicted'], squared = False)

1.6685701407225697

In [61]:
mse = sum((ads['sales_predicted'] - ads['sales'])**2) / ads.shape[0]

In [62]:
mse

2.784126314510937

In [65]:
import math
rmse=math.sqrt(mse)
rmse

1.66857014072257

In [75]:
#to implement interation terms
ads['TVXRadio'] =ads['TV'] * ads['radio']

In [76]:
ads.head()

TV  radio  newspaper  sales  sales_predicted  TVXRadio
0  230.1   37.8       69.2   22.1        20.523974   8697.78
1   44.5   39.3       45.1   10.4        12.337855   1748.85
2   17.2   45.9       69.3    9.3        12.307671    789.48
3  151.5   41.3       58.5   18.5        17.597830   6256.95
4  180.8   10.8       58.4   12.9        13.188672   1952.64

In [82]:
# prepare the variables:dependent,independent and the constant
y = ads['sales']

X_interact = ads[['TV', 'radio', 'TVXRadio']]

X_with_interact = sm.add_constant(X_interact)

# Fit the linear regression model ols means ordinatry list square
model_interact = sm.OLS(y, X_with_interact)
results_interact = model_interact.fit()

In [83]:
results_interact.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.968
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                     1963.
Date:                Mon, 19 Feb 2024   Prob (F-statistic):          6.68e-146
Time:                        15:46:03   Log-Likelihood:                -270.14
No. Observations:                 200   AIC:                             548.3
Df Residuals:                     196   BIC:                             561.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.7502      0.248     27.233      0.000       6.261       7.239
TV             0.0191      0.002     12.699      0.000       0.016       0.022
radio          0.0289      0.009      3.241      0.001       0.011       0.046
TVXRadio       0.0011   5.24e-05     20.727      0.000       0.001       0.001
==============================================================================
Omnibus:                      128.132   Durbin-Watson:                   2.224
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1183.719
Skew:                          -2.323   Prob(JB):                    9.09e-258
Kurtosis:                      13.975   Cond. No.                     1.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [64]:
happy = pd.read_csv('Happiness2019.csv')

In [12]:
happy.head()


Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   
3             4           Iceland  7.494           1.380           1.624   
4             5       Netherlands  7.488           1.396           1.522   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   
3                    1.026                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341  
3                      0.118  
4                      0.298

In [13]:
happy.columns


Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption'],
      dtype='object')

In [14]:
#how social and economic factors affect the happiness? 
#prediction vs inference
#predictive question
#descriptive question like which country is doing very good in economy by seeing the gdp per capita,
#which countries are happy?
#we can define that if their happiness score above average ,they are happy

In [15]:
happy['Score'].mean()

5.407096153846155

In [16]:
happy_countries = happy[happy['Score']>happy['Score'].mean()]
happy_countries.head(50)                        

Overall rank     Country or region  Score  GDP per capita  Social support  \
0              1               Finland  7.769           1.340           1.587   
1              2               Denmark  7.600           1.383           1.573   
2              3                Norway  7.554           1.488           1.582   
3              4               Iceland  7.494           1.380           1.624   
4              5           Netherlands  7.488           1.396           1.522   
5              6           Switzerland  7.480           1.452           1.526   
6              7                Sweden  7.343           1.387           1.487   
7              8           New Zealand  7.307           1.303           1.557   
8              9                Canada  7.278           1.365           1.505   
9             10               Austria  7.246           1.376           1.475   
10            11             Australia  7.228           1.372           1.548   
11            12            Costa Rica  7.167           1.034           1.441   
12            13                Israel  7.139           1.276           1.455   
13            14            Luxembourg  7.090           1.609           1.479   
14            15        United Kingdom  7.054           1.333           1.538   
15            16               Ireland  7.021           1.499           1.553   
16            17               Germany  6.985           1.373           1.454   
17            18               Belgium  6.923           1.356           1.504   
18            19         United States  6.892           1.433           1.457   
19            20        Czech Republic  6.852           1.269           1.487   
20            21  United Arab Emirates  6.825           1.503           1.310   
21            22                 Malta  6.726           1.300           1.520   
22            23                Mexico  6.595           1.070           1.323   
23            24                France  6.592           1.324           1.472   
24            25                Taiwan  6.446           1.368           1.430   
25            26                 Chile  6.444           1.159           1.369   
26            27             Guatemala  6.436           0.800           1.269   
27            28          Saudi Arabia  6.375           1.403           1.357   
28            29                 Qatar  6.374           1.684           1.313   
29            30                 Spain  6.354           1.286           1.484   
30            31                Panama  6.321           1.149           1.442   
31            32                Brazil  6.300           1.004           1.439   
32            33               Uruguay  6.293           1.124           1.465   
33            34             Singapore  6.262           1.572           1.463   
34            35           El Salvador  6.253           0.794           1.242   
35            36                 Italy  6.223           1.294           1.488   
36            37               Bahrain  6.199           1.362           1.368   
37            38              Slovakia  6.198           1.246           1.504   
38            39     Trinidad & Tobago  6.192           1.231           1.477   
39            40                Poland  6.182           1.206           1.438   
40            41            Uzbekistan  6.174           0.745           1.529   
41            42             Lithuania  6.149           1.238           1.515   
42            43              Colombia  6.125           0.985           1.410   
43            44              Slovenia  6.118           1.258           1.523   
44            45             Nicaragua  6.105           0.694           1.325   
45            46                Kosovo  6.100           0.882           1.232   
46            47             Argentina  6.086           1.092           1.432   
47            48               Romania  6.070           1.162           1.232   
48            49                Cyp

In [17]:
#which country is least happy?


In [18]:
happy[happy['Score']==happy['Score'].min()]


Overall rank Country or region  Score  GDP per capita  Social support  \
155           156       South Sudan  2.853           0.306           0.575   

     Healthy life expectancy  Freedom to make life choices  Generosity  \
155                    0.295                          0.01       0.202   

     Perceptions of corruption  
155                      0.091

In [19]:
happy.columns


Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption'],
      dtype='object')

In [20]:
#using linear regression for answering predictive
# prepare the variables:dependent,independent and the constant
y = happy['Score']

X = happy[['GDP per capita', 'Social support', 'Healthy life expectancy','Freedom to make life choices','Generosity','Perceptions of corruption']]

X_with_intercept = sm.add_constant(X)

# Fit the linear regression model ols means ordinatry list square
model = sm.OLS(y, X_with_intercept)
resulte = model.fit()


In [21]:
resulte.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Score   R-squared:                       0.779
Model:                            OLS   Adj. R-squared:                  0.770
Method:                 Least Squares   F-statistic:                     87.62
Date:                Mon, 19 Feb 2024   Prob (F-statistic):           2.40e-46
Time:                        14:47:02   Log-Likelihood:                -119.76
No. Observations:                 156   AIC:                             253.5
Df Residuals:                     149   BIC:                             274.9
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const                            1.7952      0.211      8.505      0.000       1.378       2.212
GDP per capita                   0.7754      0.218      3.553      0.001       0.344       1.207
Social support                   1.1242      0.237      4.745      0.000       0.656       1.592
Healthy life expectancy          1.0781      0.335      3.223      0.002       0.417       1.739
Freedom to make life choices     1.4548      0.375      3.876      0.000       0.713       2.197
Generosity                       0.4898      0.498      0.984      0.327      -0.494       1.473
Perceptions of corruption        0.9723      0.542      1.793      0.075      -0.099       2.044
==============================================================================
Omnibus:                        8.188   Durbin-Watson:                   1.648
Prob(Omnibus):                  0.017   Jarque-Bera (JB):                7.971
Skew:                          -0.498   Prob(JB):                       0.0186
Kurtosis:                       3.483   Cond. No.                         28.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [22]:
#dummy variable

In [23]:
toyata = pd.read_csv('ToyotaCorolla.csv')

In [24]:
toyata.head()

Price  Age     KM Fuel_Type  HP  MetColor  Automatic  Doors
0  13500   23  46986    Diesel  90         1          0      3
1  13750   23  72937    Diesel  90         1          0      3
2  13950   24  41711    Diesel  90         1          0      3
3  14950   26  48000    Diesel  90         0          0      3
4  13750   30  38500    Diesel  90         0          0      3

In [25]:
toyata.columns


Index(['Price', 'Age', 'KM', 'Fuel_Type', 'HP', 'MetColor', 'Automatic',
       'Doors'],
      dtype='object')

In [26]:
toyata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Price      1436 non-null   int64 
 1   Age        1436 non-null   int64 
 2   KM         1436 non-null   int64 
 3   Fuel_Type  1436 non-null   object
 4   HP         1436 non-null   int64 
 5   MetColor   1436 non-null   int64 
 6   Automatic  1436 non-null   int64 
 7   Doors      1436 non-null   int64 
dtypes: int64(7), object(1)
memory usage: 89.9+ KB


In [27]:
# prepare the variables:dependent,independent and the constant
y_car = toyata['Price']

X_car = toyata[['Age', 'KM','HP', 'MetColor', 'Automatic']]

X_with_intercept_car = sm.add_constant(X_car)

# Fit the linear regression model ols means ordinatry list square
model_car = sm.OLS(y_car, X_with_intercept_car)
results_car = model_car.fit()

In [28]:
results_car.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.813
Model:                            OLS   Adj. R-squared:                  0.812
Method:                 Least Squares   F-statistic:                     1240.
Date:                Mon, 19 Feb 2024   Prob (F-statistic):               0.00
Time:                        14:47:02   Log-Likelihood:                -12605.
No. Observations:                1436   AIC:                         2.522e+04
Df Residuals:                    1430   BIC:                         2.525e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.637e+04    355.781     45.998      0.000    1.57e+04    1.71e+04
Age         -155.4339      2.605    -59.658      0.000    -160.545    -150.323
KM            -0.0108      0.001     -7.994      0.000      -0.013      -0.008
HP            36.7672      2.943     12.494      0.000      30.994      42.540
MetColor      44.0261     89.252      0.493      0.622    -131.052     219.105
Automatic    747.9693    182.321      4.102      0.000     390.325    1105.614
==============================================================================
Omnibus:                      430.424   Durbin-Watson:                   1.349
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3263.789
Skew:                           1.183   Prob(JB):                         0.00
Kurtosis:                       9.996   Cond. No.                     6.70e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.7e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

for that two dummy variable:
    metcolor:comparing to cars with no metallic ,cars with it sellls 44 dollars more,holding everything else constant
        automatic:comparing to cars with no auto transmission,cars with it sells for 748 dollars more,holding everything else constant

In [29]:
toyata['Fuel_Type'].unique()
toyata['Fuel_Type']=toyata['Fuel_Type'].astype('category')


array(['Diesel', 'Petrol', 'CNG'], dtype=object)

In [30]:
#change the variable data from integer to category
toyata['Doors'].unique()
toyata['Doors']=toyata['Doors'].astype('category')

array([3, 5, 4, 2])

In [31]:
y_car = toyata['Price']
X_car_1 = toyata[['Age', 'KM','HP', 'MetColor', 'Automatic','Fuel_Type','Doors']]

X_car_1 = pd.get_dummies(X_car_1,columns=['Fuel_Type','Doors'],drop_first=True,dtype=int)



In [32]:
X_car_1.head()

Age     KM  HP  MetColor  Automatic  Fuel_Type_Diesel  Fuel_Type_Petrol  \
0   23  46986  90         1          0                 1                 0   
1   23  72937  90         1          0                 1                 0   
2   24  41711  90         1          0                 1                 0   
3   26  48000  90         0          0                 1                 0   
4   30  38500  90         0          0                 1                 0   

   Doors_3  Doors_4  Doors_5  
0        1        0        0  
1        1        0        0  
2        1        0        0  
3        1        0        0  
4        1        0        0

In [33]:
# Example using reindex():
#y_reindexed = y.reindex(X_car_1.index)
y_car = toyata['Price']

# Add a constant term to the independent variables
X_with_intercept_car_1 = sm.add_constant(X_car_1)

# Fit the linear regression model
model_car_1 = sm.OLS(y_car, X_with_intercept_car_1)
results_car_1 = model_car_1.fit()

# Display the summary of regression results
print(results_car_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  Price   R-squared:                       0.837
Model:                            OLS   Adj. R-squared:                  0.835
Method:                 Least Squares   F-statistic:                     729.4
Date:                Mon, 19 Feb 2024   Prob (F-statistic):               0.00
Time:                        14:47:02   Log-Likelihood:                -12506.
No. Observations:                1436   AIC:                         2.503e+04
Df Residuals:                    1425   BIC:                         2.509e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const             1.342e+04   1170.837  

4. make predictions